File: backend/routes/analytics.py
"""`/api/analytics` (`docs/API.md`, `docs/PROJECT_SPEC.md` §19)."""

from __future__ import annotations

from flask import Blueprint, g

from backend.auth.decorators import login_required
from backend.db import session as db_session
from backend.services import analytics_service
from backend.utils import success_response

analytics_bp = Blueprint("analytics", __name__)


@analytics_bp.get("/dashboard")
@login_required
def get_dashboard():
    db = db_session.get_session()
    return success_response(analytics_service.dashboard_summary(db, g.current_user.organization_id))


@analytics_bp.get("/risk-distribution")
@login_required
def get_risk_distribution():
    db = db_session.get_session()
    return success_response(analytics_service.risk_distribution(db, g.current_user.organization_id))


@analytics_bp.get("/churn-trend")
@login_required
def get_churn_trend():
    db = db_session.get_session()
    return success_response(analytics_service.churn_trend(db, g.current_user.organization_id))


@analytics_bp.get("/top-drivers")
@login_required
def get_top_drivers():
    db = db_session.get_session()
    return success_response(analytics_service.top_drivers(db, g.current_user.organization_id))


@analytics_bp.get("/segments")
@login_required
def get_segments():
    db = db_session.get_session()
    return success_response(analytics_service.segments(db, g.current_user.organization_id))

File: frontend/lib/api/analytics.ts
import { apiRequest } from "./client";
import type { GlobalShap } from "./models";
import type { RiskLevel } from "./types";

export interface ActiveModelStatus {
  available: boolean;
  model_id?: number | string | null;
  algorithm?: string | null;
  model_type?: string | null;
  version?: number;
  is_baseline?: boolean;
}

/** `backend/services/analytics_service.py::dashboard_summary`: rate/average
 * fields are `null` (not `0` or a fabricated number) until at least one
 * customer has been scored — the empty state renders on `null`, never a
 * plausible-looking placeholder (`docs/PROJECT_SPEC.md` binding rule). */
export interface DashboardSummary {
  total_customers: number;
  scored_customers: number;
  at_risk_customers: number;
  high_risk_customers: number;
  predicted_churn_rate: number | null;
  avg_churn_probability: number | null;
  active_model: ActiveModelStatus;
}

export interface RiskDistribution {
  counts: Record<RiskLevel, number>;
  scored_customers: number;
}

export interface ChurnTrendPoint {
  date: string;
  prediction_count: number;
  predicted_churn_rate: number;
  avg_churn_probability: number;
}

export interface ChurnTrend {
  points: ChurnTrendPoint[];
}

export interface TopDrivers {
  available: boolean;
  feature_importance?: GlobalShap["feature_importance"];
  disclaimer?: string;
}

export type SegmentBreakdown = Record<string, Record<RiskLevel, number>>;

export interface Segments {
  segments: Record<string, SegmentBreakdown>;
}

export const analyticsApi = {
  dashboard: () => apiRequest<DashboardSummary>("/api/analytics/dashboard"),
  riskDistribution: () => apiRequest<RiskDistribution>("/api/analytics/risk-distribution"),
  churnTrend: () => apiRequest<ChurnTrend>("/api/analytics/churn-trend"),
  topDrivers: () => apiRequest<TopDrivers>("/api/analytics/top-drivers"),
  segments: () => apiRequest<Segments>("/api/analytics/segments"),
};
